### PR TITLE
Add Paper Only TargetHitScriptEvent

### PR DESCRIPTION
**Pull request by Joo200**
Add the Event from Paper when a Target block is hit by an arrow to Denizen.

### CODE REVIEW

File: paper/src/main/java/com/denizenscript/denizen/paper/events/TargetHitScriptEvent.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    @Override
+    public ScriptEntryData getScriptEntryData() {
+        return new BukkitScriptEntryData(shooter);
+    }
+
+    @EventHandler
+    public void onProjectileHit(TargetHitEvent event) {
+        this.event = event;
+        projectile = new EntityTag(event.getEntity());
+
```
Comment (mcmonkey4eva) [inline paper/src/main/java/com/denizenscript/denizen/paper/events/TargetHitScriptEvent.java]: stray newlines

File: paper/src/main/java/com/denizenscript/denizen/paper/events/TargetHitScriptEvent.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        // Additional checks for some rare edge-cases
+        if (projectile.getLocation() == null) {
+            return;
+        }
+        if (Double.isNaN(projectile.getLocation().getDirection().normalize().getX())) {
```
Comment (mcmonkey4eva) [inline paper/src/main/java/com/denizenscript/denizen/paper/events/TargetHitScriptEvent.java]: This is clearly copypasted from the projectile hit event - is it even relevant to have here?

File: paper/src/main/java/com/denizenscript/denizen/paper/events/TargetHitScriptEvent.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import com.denizenscript.denizen.utilities.implementation.BukkitScriptEntryData;
+import com.denizenscript.denizencore.objects.ObjectTag;
+import com.denizenscript.denizencore.objects.core.ElementTag;
+import com.denizenscript.denizencore.scripts.ScriptEntryData;
+import io.papermc.paper.event.block.TargetHitEvent;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+
+
+public class TargetHitScriptEvent extends BukkitScriptEvent implements Listener {
```
Comment (mcmonkey4eva) [inline paper/src/main/java/com/denizenscript/denizen/paper/events/TargetHitScriptEvent.java]: classname should match the Denizen name not the paper name - ie `TargetBlockHitScriptEvent`

File: paper/src/main/java/com/denizenscript/denizen/paper/events/TargetHitScriptEvent.java (hunk 2 above)
Comment (mcmonkey4eva) [reply paper/src/main/java/com/denizenscript/denizen/paper/events/TargetHitScriptEvent.java]: Relatedly - is this actually a unique event? ie is it not already covered by the existing projectile hit event?

Comment (mcmonkey4eva) [reply paper/src/main/java/com/denizenscript/denizen/paper/events/TargetHitScriptEvent.java]: Glance at source code `public class TargetHitEvent extends ProjectileHitEvent` indicates it's not unique

Comment (Joo200) [reply paper/src/main/java/com/denizenscript/denizen/paper/events/TargetHitScriptEvent.java]: I don't see a reason why this should be needed. I copied it from the ProjectileHitEvent. In paper the Location of an entity is marked with @NotNull so I removed those checks

Comment (Joo200) [reply paper/src/main/java/com/denizenscript/denizen/paper/events/TargetHitScriptEvent.java]: > Glance at source code `public class TargetHitEvent extends ProjectileHitEvent` indicates it's not unique

It has a different HandlerList so it's a different event

File: paper/src/main/java/com/denizenscript/denizen/paper/events/TargetBlockHitScriptEvent.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+    public TargetHitEvent event;
+    public LocationTag hitBlock;
+    public EntityTag projectile;
+    public EntityTag shooter;
+    public ElementTag strength;
+
+    @Override
+    public boolean couldMatch(ScriptPath path) {
+        return path.eventLower.matches("targetblock hit");
```
Comment (mcmonkey4eva) [inline paper/src/main/java/com/denizenscript/denizen/paper/events/TargetBlockHitScriptEvent.java]: This is the legacy couldMatch handling, should use the modern register handling

File: paper/src/main/java/com/denizenscript/denizen/paper/events/TargetBlockHitScriptEvent.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+    @Override
+    public ScriptEntryData getScriptEntryData() {
+        return new BukkitScriptEntryData(shooter);
+    }
+
+    @EventHandler
+    public void onProjectileHit(TargetHitEvent event) {
+        this.event = event;
+        projectile = new EntityTag(event.getEntity());
```
Comment (mcmonkey4eva) [inline paper/src/main/java/com/denizenscript/denizen/paper/events/TargetBlockHitScriptEvent.java:78]: none of these need to be read in advance, they should just be read in the getContext call

EDIT: oh, wait, some of these do -- just not all

File: paper/src/main/java/com/denizenscript/denizen/paper/events/TargetBlockHitScriptEvent.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        hitBlock = new LocationTag(event.getHitBlock().getLocation());
+        shooter = projectile.getShooter();
+        strength = new ElementTag(event.getSignalStrength());
```
Comment (mcmonkey4eva) [inline paper/src/main/java/com/denizenscript/denizen/paper/events/TargetBlockHitScriptEvent.java]: This doesn't need to be read in advance, should just be in the getContext call

Comment (Joo200) [reply paper/src/main/java/com/denizenscript/denizen/paper/events/TargetBlockHitScriptEvent.java]: I moved the Strengh attribute to the getContext call. The other Tags are used in other methods. If you want to move them too I would have to create new Tags every call.

**Comment by Joo200** [comment]
Sorry for the delay of this PR.

I rebased this PR and applied all PR comments

**Comment by mcmonkey4eva** [comment]
Why the rebase?

File: paper/src/main/java/com/denizenscript/denizen/paper/events/TargetBlockHitScriptEvent.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    // @Group Paper
+    //
+    // @Plugin Paper
+    //
+    // @Triggers when a target block is hit by a projectile such as an arrow.
+    //
+    // @Context
+    // <context.projectile> returns an EntityTag of the projectile.
+    // <context.shooter> returns a EntityTag of the shooter of the projectile.
+    // <context.hit_block> returns a LocationTag of the block
```
Comment (mcmonkey4eva) [inline paper/src/main/java/com/denizenscript/denizen/paper/events/TargetBlockHitScriptEvent.java]: Needs periods at end of line

Comment (mcmonkey4eva) [reply paper/src/main/java/com/denizenscript/denizen/paper/events/TargetBlockHitScriptEvent.java]: ... oh and, for that matter, the rest of the text lol.

**Comment by Joo200** [comment]
> Why the rebase?

Force of habit

File: paper/src/main/java/com/denizenscript/denizen/paper/events/TargetBlockHitScriptEvent.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    //
+    // @Plugin Paper
+    //
+    // @Triggers when a target block is hit by a projectile such as an arrow.
+    //
+    // @Context
+    // <context.projectile> returns an EntityTag of the projectile.
+    // <context.shooter> returns a EntityTag of the shooter of the projectile.
+    // <context.hit_block> returns a LocationTag of the block.
+    // <context.hit_face> returns a LocationTag of the face.
```
Comment (mcmonkey4eva) [inline paper/src/main/java/com/denizenscript/denizen/paper/events/TargetBlockHitScriptEvent.java]: These were copied from the `projectile hits` event, but cut short -- should probably just restore the full text from the other events docs into here

**Review by mcmonkey4eva — DISMISSED** [review]

File: paper/src/main/java/com/denizenscript/denizen/paper/events/TargetBlockHitScriptEvent.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+import io.papermc.paper.event.block.TargetHitEvent;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+
+
+public class TargetBlockHitScriptEvent extends BukkitScriptEvent implements Listener {
+
+    // <--[event]
+    // @Events
+    // targetblock hit
```
Comment (tal5) [inline paper/src/main/java/com/denizenscript/denizen/paper/events/TargetBlockHitScriptEvent.java]: Event name should have a space (`target block hit`)

**Comment by mcmonkey4eva** [comment]
Please stop force-pushing, it makes the PR hard to track

File: paper/src/main/java/com/denizenscript/denizen/paper/events/TargetBlockHitScriptEvent.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    @Override
+    public ObjectTag getContext(String name) {
+        return switch (name) {
+            case "projectile" -> projectile.getDenizenObject();
+            case "hit_block" -> hitBlock;
+            case "hit_face" -> event.getHitBlockFace() != null ? new LocationTag(event.getHitBlockFace().getDirection()) : null;
+            case "shooter" -> shooter != null ? shooter.getDenizenObject() : null;
+            case "strength" ->  new ElementTag(event.getSignalStrength());
```
Comment (mcmonkey4eva) [inline paper/src/main/java/com/denizenscript/denizen/paper/events/TargetBlockHitScriptEvent.java]: oh, tiny minor thing: double space here

Comment (Joo200) [reply paper/src/main/java/com/denizenscript/denizen/paper/events/TargetBlockHitScriptEvent.java]:
```suggestion
            case "strength" -> new ElementTag(event.getSignalStrength());
```

**Comment by Joo200** [comment]
> Please stop force-pushing, it makes the PR hard to track

Sorry, wanted to test this on the latest version to verify nothing is broken in 1.20.6

**Review by mcmonkey4eva — APPROVED** [review]

**Review by tal5 — APPROVED** [review]

**Pull request merged**